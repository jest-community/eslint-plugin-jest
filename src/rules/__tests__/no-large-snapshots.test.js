import { RuleTester } from 'eslint';
import { parse } from 'babel-eslint';
import rule from '../no-large-snapshots';

const noLargeSnapshots = rule.create;

const ruleTester = new RuleTester({
  parserOptions: {
    ecmaVersion: 2015,
  },
});

// lines - 1 to account for the starting newline we always add.
const generateSnapshotNode = ({ lines, title = 'a big component 1' }) =>
  parse(`exports[\`${title}\`] = \`\n${'line\n'.repeat(lines - 1)}\`;`).body[0];

ruleTester.run('no-large-snapshots', rule, {
  valid: [
    {
      filename: 'mock.js',
      code: `expect(something).toMatchInlineSnapshot(\`\n${'line\n'.repeat(
        2,
      )}\`);`,
    },
    {
      filename: 'mock.js',
      code: `expect(something).toThrowErrorMatchingInlineSnapshot(\`\n${'line\n'.repeat(
        2,
      )}\`);`,
    },
  ],
  invalid: [
    {
      filename: 'mock.js',
      code: `expect(something).toMatchInlineSnapshot(\`\n${'line\n'.repeat(
        50,
      )}\`);`,
      errors: [
        {
          messageId: 'tooLongSnapshots',
          data: { lineLimit: 50, lineCount: 51 },
        },
      ],
    },
    {
      filename: 'mock.js',
      code: `expect(something).toThrowErrorMatchingInlineSnapshot(\`\n${'line\n'.repeat(
        50,
      )}\`);`,
      errors: [
        {
          messageId: 'tooLongSnapshots',
          data: { lineLimit: 50, lineCount: 51 },
        },
      ],
    },
  ],
});

// was not able to use https://eslint.org/docs/developer-guide/nodejs-api#ruletester for these because there is no way to configure RuleTester to run non .js files
describe('no-large-snapshots', () => {
  it('should return an empty object for non snapshot files', () => {
    const mockContext = {
      getFilename: () => 'mock-component.jsx',
      options: [],
    };
    const result = noLargeSnapshots(mockContext);

    expect(result).toEqual({});
  });

  it('should return an object with an ExpressionStatement function for snapshot files', () => {
    const mockContext = {
      getFilename: () => '/mock-component.jsx.snap',
      options: [],
    };

    const result = noLargeSnapshots(mockContext);

    expect(result).toMatchObject({
      ExpressionStatement: expect.any(Function),
    });
  });

  describe('ExpressionStatement function', () => {
    it('should report if node has more than 50 lines of code and no sizeThreshold option is passed', () => {
      const mockReport = jest.fn();
      const mockContext = {
        getFilename: () => '/mock-component.jsx.snap',
        options: [],
        report: mockReport,
      };
      const mockNode = {
        loc: {
          start: {
            line: 1,
          },
          end: {
            line: 53,
          },
        },
      };
      noLargeSnapshots(mockContext).ExpressionStatement(mockNode);

      expect(mockReport).toHaveBeenCalledTimes(1);
      expect(mockReport.mock.calls[0]).toMatchSnapshot();
    });

    it('should report if node has more lines of code than number given in sizeThreshold option', () => {
      const mockReport = jest.fn();
      const mockContext = {
        getFilename: () => '/mock-component.jsx.snap',
        options: [{ maxSize: 70 }],
        report: mockReport,
      };
      const mockNode = {
        loc: {
          start: {
            line: 20,
          },
          end: {
            line: 103,
          },
        },
      };
      noLargeSnapshots(mockContext).ExpressionStatement(mockNode);

      expect(mockReport).toHaveBeenCalledTimes(1);
      expect(mockReport.mock.calls[0]).toMatchSnapshot();
    });

    it('should report if maxSize is zero', () => {
      const mockReport = jest.fn();
      const mockContext = {
        getFilename: () => '/mock-component.jsx.snap',
        options: [{ maxSize: 0 }],
        report: mockReport,
      };
      const mockNode = {
        loc: {
          start: {
            line: 1,
          },
          end: {
            line: 2,
          },
        },
      };
      noLargeSnapshots(mockContext).ExpressionStatement(mockNode);

      expect(mockReport).toHaveBeenCalledTimes(1);
      expect(mockReport.mock.calls[0]).toMatchSnapshot();
    });

    it('should not report if node has fewer lines of code than limit', () => {
      const mockReport = jest.fn();
      const mockContext = {
        getFilename: () => '/mock-component.jsx.snap',
        options: [],
        report: mockReport,
      };
      const mockNode = {
        loc: {
          start: {
            line: 1,
          },
          end: {
            line: 18,
          },
        },
      };
      noLargeSnapshots(mockContext).ExpressionStatement(mockNode);

      expect(mockReport).not.toHaveBeenCalled();
    });

    it('should not report whitelisted large snapshots', () => {
      const mockReport = jest.fn();
      const mockContext = {
        getFilename: () => '/mock-component.jsx.snap',
        options: [
          {
            whitelistedSnapshots: {
              '/mock-component.jsx.snap': ['a big component 1'],
            },
          },
        ],
        report: mockReport,
      };

      const snapshotNode = generateSnapshotNode({ lines: 58 });

      noLargeSnapshots(mockContext).ExpressionStatement(snapshotNode);

      expect(mockReport).not.toHaveBeenCalled();
    });

    it('should report if file is not whitelisted', () => {
      const mockReport = jest.fn();
      const mockContext = {
        getFilename: () => '/mock-component.jsx.snap',
        options: [
          {
            whitelistedSnapshots: {
              '/other-mock-component.jsx.snap': [/a big component \d+/],
            },
          },
        ],
        report: mockReport,
      };

      const snapshotNode = generateSnapshotNode({ lines: 58 });

      noLargeSnapshots(mockContext).ExpressionStatement(snapshotNode);

      expect(mockReport).toHaveBeenCalledTimes(1);
      expect(mockReport.mock.calls[0]).toMatchSnapshot();
    });

    it('should not report whitelisted large snapshots based on regexp', () => {
      const mockReport = jest.fn();
      const mockContext = {
        getFilename: () => '/mock-component.jsx.snap',
        options: [
          {
            whitelistedSnapshots: {
              '/mock-component.jsx.snap': [/a big component \d+/],
            },
          },
        ],
        report: mockReport,
      };

      const snapshotNode = generateSnapshotNode({ lines: 58 });

      noLargeSnapshots(mockContext).ExpressionStatement(snapshotNode);

      expect(mockReport).not.toHaveBeenCalled();

      const otherSnapshotNode = generateSnapshotNode({
        lines: 58,
        title: 'a big component with text',
      });

      noLargeSnapshots(mockContext).ExpressionStatement(otherSnapshotNode);

      expect(mockReport).toHaveBeenCalledTimes(1);
      expect(mockReport.mock.calls[0]).toMatchSnapshot();
    });

    it('should throw exeption if relative paths are passed as whitelist keys', () => {
      const mockReport = jest.fn();
      const mockContext = {
        getFilename: () => '/mock-component.jsx.snap',
        options: [
          {
            whitelistedSnapshots: {
              'mock-component.jsx.snap': [/a big component \d+/],
            },
          },
        ],
        report: mockReport,
      };

      const snapshotNode = generateSnapshotNode({ lines: 58 });

      expect(() =>
        noLargeSnapshots(mockContext).ExpressionStatement(snapshotNode),
      ).toThrow(
        'All paths for whitelistedSnapshots must be absolute. You can use JS config and `path.resolve`',
      );
    });

    it('should not throw exeption if absolute paths are passed as whitelist keys', () => {
      const mockReport = jest.fn();
      const mockContext = {
        getFilename: () => '/mock-component.jsx.snap',
        options: [
          {
            whitelistedSnapshots: {
              '/mock-component.jsx.snap': [/a big component \d+/],
            },
          },
        ],
        report: mockReport,
      };

      const snapshotNode = generateSnapshotNode({ lines: 58 });

      expect(() =>
        noLargeSnapshots(mockContext).ExpressionStatement(snapshotNode),
      ).not.toThrow();
    });
  });
});
