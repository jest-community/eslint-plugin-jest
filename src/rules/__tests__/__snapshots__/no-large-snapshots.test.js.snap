// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no-large-snapshots ExpressionStatement function should not report whitelisted large snapshots based on regexp 1`] = `
Array [
  Object {
    "data": Object {
      "lineCount": 58,
      "lineLimit": 50,
    },
    "messageId": "tooLongSnapshots",
    "node": Node {
      "_babelType": "ExpressionStatement",
      "end": 328,
      "expression": Node {
        "_babelType": "AssignmentExpression",
        "end": 327,
        "left": Node {
          "_babelType": "MemberExpression",
          "computed": true,
          "end": 36,
          "loc": SourceLocation {
            "end": Position {
              "column": 36,
              "line": 1,
            },
            "start": Position {
              "column": 0,
              "line": 1,
            },
          },
          "object": Node {
            "_babelType": "Identifier",
            "end": 7,
            "loc": SourceLocation {
              "end": Position {
                "column": 7,
                "line": 1,
              },
              "identifierName": "exports",
              "start": Position {
                "column": 0,
                "line": 1,
              },
            },
            "name": "exports",
            "range": Array [
              0,
              7,
            ],
            "start": 0,
            "type": "Identifier",
          },
          "property": Node {
            "_babelType": "TemplateLiteral",
            "end": 35,
            "expressions": Array [],
            "loc": SourceLocation {
              "end": Position {
                "column": 35,
                "line": 1,
              },
              "start": Position {
                "column": 8,
                "line": 1,
              },
            },
            "quasis": Array [
              Node {
                "_babelType": "TemplateElement",
                "end": 34,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 35,
                    "line": 1,
                  },
                  "start": Position {
                    "column": 8,
                    "line": 1,
                  },
                },
                "range": Array [
                  8,
                  35,
                ],
                "start": 9,
                "tail": true,
                "type": "TemplateElement",
                "value": Object {
                  "cooked": "a big component with text",
                  "raw": "a big component with text",
                },
              },
            ],
            "range": Array [
              8,
              35,
            ],
            "start": 8,
            "type": "TemplateLiteral",
          },
          "range": Array [
            0,
            36,
          ],
          "start": 0,
          "type": "MemberExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 1,
            "line": 59,
          },
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "operator": "=",
        "range": Array [
          0,
          327,
        ],
        "right": Node {
          "_babelType": "TemplateLiteral",
          "end": 327,
          "expressions": Array [],
          "loc": SourceLocation {
            "end": Position {
              "column": 1,
              "line": 59,
            },
            "start": Position {
              "column": 39,
              "line": 1,
            },
          },
          "quasis": Array [
            Node {
              "_babelType": "TemplateElement",
              "end": 326,
              "loc": SourceLocation {
                "end": Position {
                  "column": 1,
                  "line": 59,
                },
                "start": Position {
                  "column": 39,
                  "line": 1,
                },
              },
              "range": Array [
                39,
                327,
              ],
              "start": 40,
              "tail": true,
              "type": "TemplateElement",
              "value": Object {
                "cooked": "
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
",
                "raw": "
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
",
              },
            },
          ],
          "range": Array [
            39,
            327,
          ],
          "start": 39,
          "type": "TemplateLiteral",
        },
        "start": 0,
        "type": "AssignmentExpression",
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 2,
          "line": 59,
        },
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "range": Array [
        0,
        328,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  },
]
`;

exports[`no-large-snapshots ExpressionStatement function should report if file is not whitelisted 1`] = `
Array [
  Object {
    "data": Object {
      "lineCount": 58,
      "lineLimit": 50,
    },
    "messageId": "tooLongSnapshots",
    "node": Node {
      "_babelType": "ExpressionStatement",
      "end": 320,
      "expression": Node {
        "_babelType": "AssignmentExpression",
        "end": 319,
        "left": Node {
          "_babelType": "MemberExpression",
          "computed": true,
          "end": 28,
          "loc": SourceLocation {
            "end": Position {
              "column": 28,
              "line": 1,
            },
            "start": Position {
              "column": 0,
              "line": 1,
            },
          },
          "object": Node {
            "_babelType": "Identifier",
            "end": 7,
            "loc": SourceLocation {
              "end": Position {
                "column": 7,
                "line": 1,
              },
              "identifierName": "exports",
              "start": Position {
                "column": 0,
                "line": 1,
              },
            },
            "name": "exports",
            "range": Array [
              0,
              7,
            ],
            "start": 0,
            "type": "Identifier",
          },
          "property": Node {
            "_babelType": "TemplateLiteral",
            "end": 27,
            "expressions": Array [],
            "loc": SourceLocation {
              "end": Position {
                "column": 27,
                "line": 1,
              },
              "start": Position {
                "column": 8,
                "line": 1,
              },
            },
            "quasis": Array [
              Node {
                "_babelType": "TemplateElement",
                "end": 26,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 27,
                    "line": 1,
                  },
                  "start": Position {
                    "column": 8,
                    "line": 1,
                  },
                },
                "range": Array [
                  8,
                  27,
                ],
                "start": 9,
                "tail": true,
                "type": "TemplateElement",
                "value": Object {
                  "cooked": "a big component 1",
                  "raw": "a big component 1",
                },
              },
            ],
            "range": Array [
              8,
              27,
            ],
            "start": 8,
            "type": "TemplateLiteral",
          },
          "range": Array [
            0,
            28,
          ],
          "start": 0,
          "type": "MemberExpression",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 1,
            "line": 59,
          },
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "operator": "=",
        "range": Array [
          0,
          319,
        ],
        "right": Node {
          "_babelType": "TemplateLiteral",
          "end": 319,
          "expressions": Array [],
          "loc": SourceLocation {
            "end": Position {
              "column": 1,
              "line": 59,
            },
            "start": Position {
              "column": 31,
              "line": 1,
            },
          },
          "quasis": Array [
            Node {
              "_babelType": "TemplateElement",
              "end": 318,
              "loc": SourceLocation {
                "end": Position {
                  "column": 1,
                  "line": 59,
                },
                "start": Position {
                  "column": 31,
                  "line": 1,
                },
              },
              "range": Array [
                31,
                319,
              ],
              "start": 32,
              "tail": true,
              "type": "TemplateElement",
              "value": Object {
                "cooked": "
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
",
                "raw": "
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
line
",
              },
            },
          ],
          "range": Array [
            31,
            319,
          ],
          "start": 31,
          "type": "TemplateLiteral",
        },
        "start": 0,
        "type": "AssignmentExpression",
      },
      "loc": SourceLocation {
        "end": Position {
          "column": 2,
          "line": 59,
        },
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "range": Array [
        0,
        320,
      ],
      "start": 0,
      "type": "ExpressionStatement",
    },
  },
]
`;

exports[`no-large-snapshots ExpressionStatement function should report if maxSize is zero 1`] = `
Array [
  Object {
    "data": Object {
      "lineCount": 1,
      "lineLimit": 0,
    },
    "messageId": "noSnapshot",
    "node": Object {
      "loc": Object {
        "end": Object {
          "line": 2,
        },
        "start": Object {
          "line": 1,
        },
      },
    },
  },
]
`;

exports[`no-large-snapshots ExpressionStatement function should report if node has more lines of code than number given in sizeThreshold option 1`] = `
Array [
  Object {
    "data": Object {
      "lineCount": 83,
      "lineLimit": 70,
    },
    "messageId": "tooLongSnapshots",
    "node": Object {
      "loc": Object {
        "end": Object {
          "line": 103,
        },
        "start": Object {
          "line": 20,
        },
      },
    },
  },
]
`;

exports[`no-large-snapshots ExpressionStatement function should report if node has more than 50 lines of code and no sizeThreshold option is passed 1`] = `
Array [
  Object {
    "data": Object {
      "lineCount": 52,
      "lineLimit": 50,
    },
    "messageId": "tooLongSnapshots",
    "node": Object {
      "loc": Object {
        "end": Object {
          "line": 53,
        },
        "start": Object {
          "line": 1,
        },
      },
    },
  },
]
`;
